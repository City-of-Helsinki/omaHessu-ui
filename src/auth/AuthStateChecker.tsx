import React, { useMemo } from 'react';

import { useLoginStateListener } from './LoginContext';
import { LoginClientStateChange } from './login-client';

const AuthStateChecker = (): React.ReactElement | null => {
  const onStateListener = useMemo(
    () => (obj: LoginClientStateChange) => {
      console.log('onStateListener', obj);
    },
    []
  );
  const { state, previousState, error } = useLoginStateListener(
    onStateListener
  );
  return (
    <div>
      <p>State:{state}</p>
      <p>PreviousState:{previousState}</p>
      <p>Error:{error ? error.type : 'none'}</p>
    </div>
  );
};

export default AuthStateChecker;
