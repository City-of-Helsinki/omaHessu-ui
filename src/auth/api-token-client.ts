import to from 'await-to-js';
import { User } from 'oidc-client-ts';

import { createFetchCanceller } from '../common/helpers/fetchCanceller';
import retryPollingUntilSuccessful from './http-poller-with-promises';
import LoginClientError from './login-client-error';

export type TokenData = Record<string, string>;
export type ApiTokenClient = {
  fetch: (user: User) => Promise<TokenData>;
  getToken: (name: string) => string | undefined;
  getTokens: () => TokenData | null;
  clear: () => void;
};

export type ApiTokenClientProps = {
  url: string;
  maxRetries?: number;
  retryInterval?: number;
};

export type FetchApiTokenOptions = {
  url: string;
  accessToken: string;
  signal?: AbortSignal;
  maxRetries?: number;
  retryInterval?: number;
};

export const API_TOKEN_SESSION_STORAGE_KEY = 'api_token_key';

async function fetchApiToken(
  options: FetchApiTokenOptions
): Promise<TokenData | LoginClientError> {
  const {
    url,
    signal,
    accessToken,
    maxRetries = 4,
    retryInterval = 500,
  } = options;
  const myHeaders = new Headers();
  myHeaders.append('Authorization', `Bearer ${accessToken}`);

  const requestOptions = {
    method: 'POST',
    headers: myHeaders,
    signal,
  };

  const pollFunction = () => fetch(url, requestOptions);

  const [fetchError, fetchResponse] = await to(
    retryPollingUntilSuccessful({
      pollFunction,
      pollIntervalInMs: retryInterval,
      maxRetries,
    })
  );

  if (fetchError || !fetchResponse) {
    return new LoginClientError(
      'Network or CORS error occured',
      'API_TOKEN_NETWORK_OR_CORS_ERROR',
      fetchError
    );
  }
  if (!fetchResponse.ok) {
    const message = await fetchResponse.text();
    return new LoginClientError(
      `${message}.${fetchResponse.statusText} Status:${fetchResponse.status}`,
      'API_TOKENS_FAILED',
      fetchError
    );
  }
  const [parseError, json] = await to(fetchResponse.json());
  if (parseError) {
    return new LoginClientError(
      parseError.message,
      'INVALID_API_TOKENS',
      parseError
    );
  }
  return json as TokenData;
}

export default function createApiTokenClient(
  props: ApiTokenClientProps
): ApiTokenClient {
  const { url, maxRetries, retryInterval } = props;
  const fetchCanceller = createFetchCanceller();

  const getStoredTokens = (): TokenData | null => {
    const tokensString = sessionStorage.getItem(API_TOKEN_SESSION_STORAGE_KEY);
    try {
      return tokensString ? JSON.parse(tokensString) : null;
    } catch (e) {
      return null;
    }
  };

  const clearStoredTokens = (): void =>
    sessionStorage.removeItem(API_TOKEN_SESSION_STORAGE_KEY);

  const setStoredTokens = (tokenObj: TokenData): void => {
    sessionStorage.setItem(
      API_TOKEN_SESSION_STORAGE_KEY,
      JSON.stringify(tokenObj)
    );
  };
  let tokens: TokenData | null = getStoredTokens() || null;
  return {
    fetch: async user => {
      fetchCanceller.cancel();
      const { access_token: accessToken } = user;
      const result = await fetchApiToken({
        url,
        accessToken,
        signal: fetchCanceller.getSignal(),
        maxRetries,
        retryInterval,
      });
      clearStoredTokens();
      if (result instanceof Error) {
        return Promise.reject(result);
      }
      tokens = { ...result };
      setStoredTokens(tokens);
      return Promise.resolve(tokens);
    },
    getToken: name => (tokens ? tokens[name] : undefined),
    getTokens: () => tokens,
    clear: () => {
      fetchCanceller.cancel();
      clearStoredTokens();
      tokens = null;
    },
  };
}
