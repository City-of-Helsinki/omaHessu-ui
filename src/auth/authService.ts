import {
  UserManager,
  User,
  UserManagerSettings,
  Log,
  WebStorageStateStore,
} from 'oidc-client-ts';
import jwtDecode from 'jwt-decode';
import * as Sentry from '@sentry/browser';
import HttpStatusCode from 'http-status-typed';
import i18n from 'i18next';
import to from 'await-to-js';

import pickProfileApiToken from './pickProfileApiToken';
import createHttpPoller, { HttpPoller } from './http-poller';
import retryPollingUntilSuccessful from './http-poller-with-promises';

const origin = window.location.origin;
export const API_TOKEN = 'apiToken';

function isUserExpired(user?: Partial<User> | null): boolean {
  if (!user) {
    return true;
  }
  if (user.expired !== undefined) {
    return user.expired;
  }
  const expiresAtInSeconds = user.expires_at;

  if (expiresAtInSeconds) {
    return expiresAtInSeconds - Date.now() / 1000 <= 0;
  }
  return true;
}

export class AuthService {
  userManager: UserManager;
  userSessionValidityPoller: HttpPoller;
  private _isProcessingLogin = false;
  constructor() {
    const settings: UserManagerSettings = {
      automaticSilentRenew: true,
      userStore: new WebStorageStateStore({ store: window.sessionStorage }),
      authority: window._env_.REACT_APP_OIDC_AUTHORITY,
      client_id: window._env_.REACT_APP_OIDC_CLIENT_ID,
      redirect_uri: `${origin}/callback`,
      silent_redirect_uri: `${origin}/silent_renew.html`,
      response_type: window._env_.REACT_APP_OIDC_RESPONSE_TYPE,
      scope: window._env_.REACT_APP_OIDC_SCOPE,
      post_logout_redirect_uri: `${origin}/`,
      monitorSession: true,
      includeIdTokenInSilentRenew: true,
      validateSubOnSilentRenew: false,
      loadUserInfo: true,
      // This calculates to 1 minute, good for debugging:
      // eslint-disable-next-line max-len
      // https://github.com/City-of-Helsinki/kukkuu-ui/blob/8029ed64c3d0496fa87fa57837c73520e8cbe37f/src/domain/auth/userManager.ts#L18
      //accessTokenExpiringNotificationTimeInSeconds: 59.65 * 60,
    };

    // Show oidc debugging info in the console only while developing
    if (window._env_.NODE_ENV === 'development') {
      Log.setLogger(console);
      Log.setLevel(Log.INFO);
    }

    // User Manager instance
    this.userManager = new UserManager(settings);

    // Public methods
    this.getUser = this.getUser.bind(this);
    this.getToken = this.getToken.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.login = this.login.bind(this);
    this.endLogin = this.endLogin.bind(this);
    this.renewToken = this.renewToken.bind(this);
    this.logout = this.logout.bind(this);

    const userInfoFetchFunction = async (): Promise<Response | undefined> => {
      const uri = await this.userManager.metadataService.getUserInfoEndpoint();
      const user = await this.getUser();
      const accessToken = user && user.access_token;
      if (!accessToken) {
        return Promise.reject(new Error('Access token not set'));
      }
      const headers = new Headers();
      headers.append('authorization', `Bearer ${accessToken}`);

      return fetch(uri, {
        method: 'GET',
        headers,
      });
    };

    this.userSessionValidityPoller = createHttpPoller({
      pollFunction: userInfoFetchFunction,
      shouldPoll: () => this.isAuthenticated(),
      onError: returnedHttpStatus => {
        if (
          this.isAuthenticated() &&
          returnedHttpStatus &&
          (returnedHttpStatus === HttpStatusCode.FORBIDDEN ||
            returnedHttpStatus === HttpStatusCode.UNAUTHORIZED)
        ) {
          this.logout();
          return { keepPolling: false };
        }
        return { keepPolling: this.isAuthenticated() };
      },
    });

    // Events
    this.userManager.events.addAccessTokenExpired(() => {
      this.logout();
    });

    this.userManager.events.addUserSignedOut(async () => {
      await this.cleanUpUserSessionAndApiTokens();
    });

    // This is called by userManager while processing endLogin()
    // and when silent renew is complete
    // endLogin() also calls fetchApiToken. Multiple calls are prevented with _isProcessingLogin
    this.userManager.events.addUserLoaded(async user => {
      if (!this._isProcessingLogin && this.isAuthenticatedUser(user)) {
        const [, response] = await to(
          retryPollingUntilSuccessful({
            pollFunction: async () => this.fetchApiToken(user),
            pollIntervalInMs: 500,
            maxRetries: 4,
          })
        );
        if (response) {
          await this.parseAndStoreApiToken(response);
        } else {
          this.logout();
        }
      }
      return Promise.resolve();
    });

    this.userManager.events.addUserUnloaded(async () => {
      await this.cleanUpUserSessionAndApiTokens();
    });
  }

  public async getUser(): Promise<User | null> {
    const user = await this.userManager.getUser();
    if (user) {
      return Promise.resolve(this.addAmrToUserProfile(user));
    }

    return Promise.resolve(null);
  }

  public addAmrToUserProfile(user: User): User {
    if (
      user &&
      user.profile.amr &&
      Array.isArray(user.profile.amr) &&
      user.profile.amr.length
    ) {
      return user;
    }
    if (!user || !user.id_token) {
      return user;
    }

    const decodedToken = jwtDecode<Record<string, string>>(user.id_token);
    const amr = decodedToken.amr;
    if (!amr) {
      return user;
    }
    user.profile.amr = Array.isArray(amr) ? amr : [amr];
    return user;
  }
  // It is assumed that user and api tokens are removed hand in hand.
  // If user tokens exists, api tokens must also exist.
  // That is why api tokens are not checked here.
  public async getAuthenticatedUser(): Promise<User | null> {
    const user = await this.getUser();
    if (!this.isAuthenticatedUser(user)) {
      return Promise.reject(null);
    }
    this.userSessionValidityPoller.start();
    return Promise.resolve(user);
  }

  public getToken(): string | null {
    return sessionStorage.getItem(API_TOKEN);
  }

  public isAuthenticatedUser(user?: User | null): boolean {
    return !!user && !isUserExpired(user) && !!user.access_token;
  }

  public isAuthenticated(): boolean {
    const userKey = `oidc.user:${window._env_.REACT_APP_OIDC_AUTHORITY}:${window._env_.REACT_APP_OIDC_CLIENT_ID}`;
    const oidcStorage = sessionStorage.getItem(userKey);
    const apiTokens = this.getToken();
    const parsedUser = oidcStorage && JSON.parse(oidcStorage);
    return this.isAuthenticatedUser(parsedUser) && !!apiTokens;
  }

  public async login(path = '/'): Promise<void> {
    let success = true;
    await this.userManager
      .signinRedirect({
        extraQueryParams: { ui_locales: i18n.language },
        state: { path },
      })
      .catch(error => {
        success = false;
        if (error.message !== 'Network Error') {
          Sentry.captureException(error);
        }
      });
    return success ? Promise.resolve() : Promise.reject();
  }

  public async endLogin(): Promise<User> {
    this._isProcessingLogin = true;
    const user = await this.userManager.signinRedirectCallback();
    if (!this.isAuthenticatedUser(user)) {
      return Promise.reject(new Error('Login failed - no valid user returned'));
    }
    const apiTokenSuccess = await this.fetchAndStoreApiToken(user);
    if (!apiTokenSuccess) {
      await this.userManager.removeUser();
      return Promise.reject(
        new Error('Api token error - no valid api token returned')
      );
    }
    this._isProcessingLogin = false;
    this.userSessionValidityPoller.start();
    return user;
  }

  public renewToken(): Promise<User | null> {
    return this.userManager.signinSilent();
  }

  public async logout(): Promise<void> {
    await this.userManager.signoutRedirect({
      extraQueryParams: { ui_locales: i18n.language },
    });
  }

  async fetchApiToken(user: User): Promise<Response | undefined> {
    const url = `${window._env_.REACT_APP_OIDC_AUTHORITY}api-tokens/`;
    return fetch(url, {
      headers: {
        authorization: `bearer ${user.access_token}`,
      },
    });
  }

  async parseApiTokenResponse(response: Response): Promise<string> {
    const result = await response.json();
    return pickProfileApiToken(result);
  }

  async parseAndStoreApiToken(
    response: Response | undefined
  ): Promise<string | void> {
    if (response) {
      const apiToken = await this.parseApiTokenResponse(response);
      sessionStorage.setItem(API_TOKEN, apiToken);
      return Promise.resolve(apiToken);
    }
    return Promise.resolve();
  }

  async fetchAndStoreApiToken(user: User): Promise<boolean> {
    const [, response] = await to(this.fetchApiToken(user));
    if (!response) {
      return Promise.resolve(false);
    }
    const apiToken = await this.parseAndStoreApiToken(response);
    return Promise.resolve(!!apiToken);
  }

  async cleanUpUserSessionAndApiTokens(): Promise<void> {
    await this.userManager.clearStaleState();
    sessionStorage.removeItem(API_TOKEN);
    this.userSessionValidityPoller.stop();
  }
}

export default new AuthService();
