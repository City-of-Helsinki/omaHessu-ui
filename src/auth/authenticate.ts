import userManager from './userManager';
import store from '../redux/store';
import { apiError } from './redux';
import * as Sentry from '@sentry/browser';

export default function(): void {
  userManager.signinRedirect().catch(error => {
    if (error.message !== 'Network Error') {
      Sentry.captureException(error);
    }
    store.dispatch(apiError(error.toString()));
  });
}
