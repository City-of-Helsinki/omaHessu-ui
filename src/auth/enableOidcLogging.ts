import Oidc from 'oidc-client';

export default function() {
  Oidc.Log.logger = console;
  Oidc.Log.level = Oidc.Log.INFO;
}
