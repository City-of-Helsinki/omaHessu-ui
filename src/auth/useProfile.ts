import { User } from 'oidc-client-ts';
import React from 'react';

import useAuth from './useAuth';

export const tunnistusSuomifiAMR = 'heltunnistussuomifi';

export type AMRStatic =
  | 'github'
  | 'google'
  | 'facebook'
  | 'yletunnus'
  | 'helsinkiAccount'
  | 'tunnistusSuomifi';

export interface Profile {
  amr: NonNullable<User['profile']['amr']>;
  auth_time: number;
  email: string;
  email_verified: boolean;
  family_name: string;
  given_name: string;
  name: string;
  nickname: string;
  sub: string;
}

export interface ProfileState {
  profile: Profile | null;
  loading: boolean;
  error: Error | null;
}

function getUserProfile(user: User | null): Profile | null {
  if (user === null || user.expired !== false) {
    return null;
  }
  const profile = { ...user.profile } as Profile;
  const amr = user.profile.amr;
  if (!amr) {
    profile.amr = [];
  } else if (typeof amr === 'string') {
    profile.amr = [amr];
  }
  return profile;
}

function useProfile(): ProfileState {
  const [profile, setProfile] = React.useState<Profile | null>(null);
  const [isLoading, setIsLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<Error | null>(null);

  const { getUser: authGetUser } = useAuth();

  React.useEffect(() => {
    let ignore = false;

    function getUser() {
      setIsLoading(true);

      authGetUser()
        .then(user => {
          if (ignore) {
            return;
          }
          setProfile(getUserProfile(user));
        })
        .catch(() => {
          if (ignore) {
            return;
          }

          setError(Error('User was not found'));
        })
        .finally(() => {
          if (ignore) {
            return;
          }

          setIsLoading(false);
        });
    }

    getUser();

    return () => {
      ignore = true;
    };
  }, [authGetUser]);

  return {
    profile,
    loading: isLoading,
    error,
  };
}

export default useProfile;
