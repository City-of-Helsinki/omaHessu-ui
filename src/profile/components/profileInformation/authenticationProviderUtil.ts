import {
  Profile,
  AMRStatic,
  tunnistusSuomifiAMR,
} from '../../../auth/useProfile';
import config from '../../../config';
import { LoginMethodType, ProfileRoot } from '../../../graphql/typings';

function getAmrFromProfileData(profile: Profile | null): string | undefined {
  return profile && Array.isArray(profile.amr) ? profile.amr[0] : '';
}

export function getAmrStatic(profile: Profile | null): AMRStatic | null {
  const amr = getAmrFromProfileData(profile);

  // If amr designates helsinki account, switch the value into a static
  // value. This setup allows the amr for Helsinki account to be
  // configured, while allowing a static reference we can use for
  // translations for instance.
  if (amr === config.helsinkiAccountAMR) {
    return 'helsinkiAccount';
  }
  if (amr === tunnistusSuomifiAMR) {
    return 'tunnistusSuomifi';
  }

  if (
    amr === 'github' ||
    amr === 'google' ||
    amr === 'facebook' ||
    amr === 'yletunnus'
  ) {
    return amr;
  }

  // If amr doesn't match any of our expectations, return null.
  return null;
}

export function hasTunnistusSuomiFiAmr(profile: Profile | null): boolean {
  return getAmrFromProfileData(profile) === tunnistusSuomifiAMR;
}

export function hasHelsinkiAccountAMR(profile: Profile | null): boolean {
  return getAmrFromProfileData(profile) === config.helsinkiAccountAMR;
}

export function hasPasswordLogin(data: ProfileRoot | undefined): boolean {
  return (
    data?.myProfile?.loginMethods?.includes(LoginMethodType.PASSWORD) ?? false
  );
}
