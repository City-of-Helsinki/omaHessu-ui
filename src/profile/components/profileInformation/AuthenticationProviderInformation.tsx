import React from 'react';
import { useTranslation } from 'react-i18next';

import useProfile from '../../../auth/useProfile';
import LabeledValue from '../../../common/labeledValue/LabeledValue';
import { getAmrStatic } from './authenticationProviderUtil';
import styles from './authenticationProviderInformation.module.css';

function AuthenticationProviderInformation(): React.ReactElement | null {
  const { t } = useTranslation();
  const { profile } = useProfile();

  const amr = getAmrStatic(profile);

  if (!amr) {
    return null;
  }

  const authenticationMethodReferenceName = t(`identityProvider.${amr}`);

  return (
    <div className={styles['container']}>
      <div className={styles['label-section']}>
        <LabeledValue
          label={t('profileInformation.authenticationMethod')}
          value={authenticationMethodReferenceName}
        />
      </div>
    </div>
  );
}

export default AuthenticationProviderInformation;
