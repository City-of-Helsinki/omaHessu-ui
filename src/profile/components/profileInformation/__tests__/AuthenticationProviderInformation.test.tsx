import React from 'react';
import { render } from '@testing-library/react';

import AuthenticationProviderInformation from '../AuthenticationProviderInformation';
import { mockProfileCreator } from '../../../../common/test/userMocking';
import * as useProfile from '../../../../auth/useProfile';
import * as authenticationProviderUtil from '../authenticationProviderUtil';

const helsinkiAccountAMR = 'helsinki_tunnus-test';

vi.spyOn(useProfile, 'default').mockImplementation(
  () =>
    (({
      profile: mockProfileCreator(),
    } as unknown) as useProfile.ProfileState)
);

describe('<AuthenticationProviderInformation /> ', () => {
  const defaultProps = {};

  const getWrapper = () =>
    render(<AuthenticationProviderInformation {...defaultProps} />);

  describe('renders correctly when AMR is helsinkiAccountAMR', () => {
    beforeAll(() => {
      window._env_.REACT_APP_HELSINKI_ACCOUNT_AMR = helsinkiAccountAMR;
    });

    afterAll(() => {
      window._env_.REACT_APP_HELSINKI_ACCOUNT_AMR = 'helsinki_tunnus';
    });

    it('should render helsinki account information as expected based on config', () => {
      vi.spyOn(authenticationProviderUtil, 'getAmrStatic').mockImplementation(
        () => 'helsinkiAccount' as useProfile.AMRStatic
      );

      const { container } = getWrapper();
      expect(container).toMatchSnapshot();
    });
  });
  describe('renders correctly when AMR is tunnistusSuomifiAMR', () => {
    it('should render suomi.fi information as expected based on config', () => {
      vi.spyOn(authenticationProviderUtil, 'getAmrStatic').mockImplementation(
        () => 'tunnistusSuomifi' as useProfile.AMRStatic
      );

      const { container } = getWrapper();
      expect(container).toMatchSnapshot();
    });
  });
});
