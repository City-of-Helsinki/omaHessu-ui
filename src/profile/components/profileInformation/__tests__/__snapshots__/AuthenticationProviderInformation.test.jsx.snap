// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthenticationProviderInformation />  renders correctly when AMR is helsinkiAccountAMR should render helsinki account information as expected based on config 1`] = `
<ProfileInformationAccountManagementLink>
  <ProfileSection>
    <section
      className="profile-section"
    >
      <div
        className="content-wrapper"
      >
        <h2
          className="section-title"
        >
          profileInformation.authenticationMethod
        </h2>
        <div
          className="text-content-wrapper"
        >
          <span
            className="value"
          >
            identityProvider.helsinkiAccount
          </span>
        </div>
      </div>
    </section>
  </ProfileSection>
</ProfileInformationAccountManagementLink>
`;

exports[`<AuthenticationProviderInformation />  renders correctly when AMR is tunnistusSuomifiAMR should render suomi.fi information as expected based on config 1`] = `
<ProfileInformationAccountManagementLink>
  <ProfileSection>
    <section
      className="profile-section"
    >
      <div
        className="content-wrapper"
      >
        <h2
          className="section-title"
        >
          profileInformation.authenticationMethod
        </h2>
        <div
          className="text-content-wrapper"
        >
          <span
            className="value"
          >
            identityProvider.tunnistusSuomifi
          </span>
        </div>
      </div>
    </section>
  </ProfileSection>
</ProfileInformationAccountManagementLink>
`;
